File: data-structures/linked/singly_linked.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class SinglyLinkedList {
  constructor() {
    this.head = null;
  }

  append(data) {
    const newNode = new Node(data);
    if (this.head === null) {
      this.head = newNode;
      return;
    }
    let last = this.head;
    while (last.next !== null) {
      last = last.next;
    }
    last.next = newNode;
  }

  printList() {
    let current = this.head;
    let result = '';
    while (current !== null) {
      result += current.data + ' -> ';
      current = current.next;
    }
    console.log(result + 'NULL');
  }
}

const list = new SinglyLinkedList();
list.append(1);
list.append(2);
list.append(3);
list.printList();  // 输出：1 -> 2 -> 3 -> NULL

/*
jarry@MacBook-Pro linked % node singly_linked.js
1 -> 2 -> 3 -> NULL
*/

File: data-structures/linked/singly_linked.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class SinglyLinkedList:
    def __init__(self):
        self.head = None

    def append(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        last = self.head
        while last.next:
            last = last.next
        last.next = new_node

    def print_list(self):
        current = self.head
        while current:
            print(f"{current.data} -> ", end="")
            current = current.next
        print("NULL")

# 使用示例
linked_list = SinglyLinkedList()
linked_list.append(1)
linked_list.append(2)
linked_list.append(3)
linked_list.print_list()  # 输出：1 -> 2 -> 3 -> NULL

"""
jarry@MacBook-Pro linked % python singly_linked.py
1 -> 2 -> 3 -> NULL
"""